src/protocol/sort/shuffle.rs

```diff
@@ -0,0 +1,99 @@
+use rand::rngs::mock::StepRng;
+use rand::seq::SliceRandom;
+
+#[allow(dead_code)]
+pub struct Shuffle {}
+
+impl Shuffle {
+    /// The Fisher–Yates shuffle is an algorithm for generating a random permutation of a finite
+    /// sequence—in plain terms, the algorithm shuffles the sequence. The algorithm effectively
+    /// puts all the elements into a hat; it continually determines the next element by randomly
+    /// drawing an element from the hat until no elements remain. The algorithm produces an unbiased
+    /// permutation: every permutation is equally likely. This algorithm takes time proportional to
+    /// the number of items being shuffled and shuffles them in place.
+    #[allow(clippy::cast_possible_truncation, dead_code)]
+    fn generate_random_sequence(
+        batchsize: usize,
+        index: u64,
+        left_permute: bool,
+        prss: &PrssSpace,
+    ) -> Vec<usize> {
+        let mut permutation: Vec<usize> = (0..batchsize).collect();
+        let rand = prss.generate_values(index.into());
+        let mut rng = if left_permute {
+            StepRng::new(rand.0 as u64, 1)
+        } else {
+            StepRng::new(rand.1 as u64, 1)
+        };
+        permutation.shuffle(&mut rng);
+        permutation
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::test_fixture::{make_participants, TestStep};
+    use rand::{rngs::mock::StepRng, Rng};
+
+    use super::Shuffle;
+
+    #[test]
+    fn random_sequence_generated() {
+        let mut rand = StepRng::new(1, 1);
+        let batchsize = 10;
+        let (p1, p2, p3) = make_participants();
+        let random_value = rand.gen::<u64>();
+        let sequence1left = Shuffle::generate_random_sequence(
+            batchsize,
+            random_value,
+            true,
+            &p1[TestStep::Shuffle],
+        );
+        let mut sequence1right = Shuffle::generate_random_sequence(
+            batchsize,
+            random_value,
+            false,
+            &p1[TestStep::Shuffle],
+        );
+
+        let sequence2left = Shuffle::generate_random_sequence(
+            batchsize,
+            random_value,
+            true,
+            &p2[TestStep::Shuffle],
+        );
+        let mut sequence2right = Shuffle::generate_random_sequence(
+            batchsize,
+            random_value,
+            false,
+            &p2[TestStep::Shuffle],
+        );
+
+        let sequence3left = Shuffle::generate_random_sequence(
+            batchsize,
+            random_value,
+            true,
+            &p3[TestStep::Shuffle],
+        );
+        let mut sequence3right = Shuffle::generate_random_sequence(
+            batchsize,
+            random_value,
+            false,
+            &p3[TestStep::Shuffle],
+        );
+
+        let expected_numbers: Vec<usize> = (0..batchsize).collect();
+        assert_eq!(sequence1right, sequence2left);
+        assert_eq!(sequence2right, sequence3left);
+        assert_eq!(sequence3right, sequence1left);
```

Comment:
I'd suggest also checking that the sequences we *don't* expect to be equal are indeed not equal. If the batchsize is large enough, the random chance of them being equal should be very small.